### PR TITLE
SAA-1412 persist offender merge event changes across affected entities

### PR DESCRIPTION
**Pull request by matt-crowson-moj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Allocation.kt [hunk 1]
```diff
@@ -404,6 +413,15 @@ data class Allocation(
     exclusions.add(exclusion)
   }
 
+  fun merge(offenderMergeDetails: OffenderMergeDetails) {
+    require(offenderMergeDetails.oldNumber == prisonerNumber) {
+      "Current prisoner number '$prisonerNumber' does not match the old prisoner merge number '${offenderMergeDetails.oldNumber}' for allocation ID '$allocationId'"
+    }
+
+    prisonerNumber = offenderMergeDetails.newNumber
+    offenderMergeDetails.newBookingId?.let { bookingId = it }
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Allocation.kt:422]: @timharrison-moj 

Speaking with Mike H he seems to thing the booking number would remain the same, so not 100% sure if we event need to support changing of the booking ID.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/OffenderMergedEventHandler.kt [hunk 2]
```diff
@@ -96,9 +87,47 @@ class OffenderMergedEventHandler(
...
-        message = "Prisoner number $oldNumber was merged to a new prisoner number $newNumber",
+        prisonCode = offenderMergeDetails.prisonCode,
+        message = "Prisoner number ${offenderMergeDetails.oldNumber} was merged to a new prisoner number ${offenderMergeDetails.newNumber}",
       ),
     )
   }
+
+  private fun mergeAllocations(mergeDetails: OffenderMergeDetails) {
+    allocationRepository.findByPrisonCodeAndPrisonerNumber(mergeDetails.prisonCode, mergeDetails.oldNumber)
+      .onEach { allocation -> allocation.merge(mergeDetails) }
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/OffenderMergedEventHandler.kt:98]: @timharrison-moj only this one does anything at the moment.  I did wonder whether it is worth feature toggling this?  Easy enough to do.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Allocation.kt (hunk 1 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Allocation.kt:422]: I wasn't sure either - but the ticket included a bit of info on it, also suggesting the booking ID would not change. However, I don't see how that can be true .. two offender records, two bookings, merged into one - so one of them must be different.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/LocalAuditRecord.kt [hunk 3]
```diff
@@ -55,4 +56,8 @@ data class LocalAuditRecord(
       message,
     )
   }
+
+  fun merge(offenderMergeDetails: OffenderMergeDetails) {
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/LocalAuditRecord.kt:60]: I suppose the idea of passing the responsibility for merging down into the entities has one drawback - that there needs to be an find, iterate through each row type loop somewhere. A quicker way might be to do this in one bulk SQL - update all attendance, all audit, all allocations?

Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/LocalAuditRecord.kt:60]: Not saying which is right or wrong - just a thought.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/AllocationTest.kt [hunk 4]
```diff
@@ -839,13 +841,39 @@ class AllocationTest {
...
+  fun `allocation offender merge fails if old number does not match`() {
+    val allocation = activityEntity().schedules().first().allocations().first().copy(allocatedPrisonerNumber = "ABC123", allocatedBookingId = 1)
+
+    assertThatThrownBy {
+      allocation.merge(OffenderMergeDetails(prisonCode = moorlandPrisonCode, oldNumber = "BDC123", newNumber = "DEF123", newBookingId = 2))
+    }.isInstanceOf(IllegalArgumentException::class.java)
+      .hasMessage("Current prisoner number 'ABC123' does not match the old prisoner merge number 'BDC123' for allocation ID '0'")
+
+    allocation.prisonerNumber isEqualTo "ABC123"
+    allocation.bookingId isEqualTo 1
```
Comment (timharrison-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/AllocationTest.kt:877]: Neat way of testing by moving the merge operation into the entity.

**Review by timharrison-moj — APPROVED** [review]
Looks good so far. 
A merge might take a while to iterate through some of the lists of entities - but is more testable this way. An alternative would be to treat this operation as an SQL data-update - to do them in one shot?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Allocation.kt (hunk 1 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Allocation.kt:422]: Yes but they always keep the latest booking ID on a merge so (in theory) we should always have the latest booking ID.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/LocalAuditRecord.kt (hunk 3 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/LocalAuditRecord.kt:60]: So thumbs up or down in terms of which way?

**Pull request closed without merging**